src/render/draw_heatmap.js

```diff
@@ -0,0 +1,137 @@
+// @flow
+
+const browser = require('../util/browser');
+const EXTENT = require('../data/extent');
+const mat4 = require('@mapbox/gl-matrix').mat4;
+const Texture = require('./texture');
+
+import type Painter from './painter';
+import type SourceCache from '../source/source_cache';
+import type HeatmapStyleLayer from '../style/style_layer/heatmap_style_layer';
+import type HeatmapBucket from '../data/bucket/heatmap_bucket';
+import type TileCoord from '../source/tile_coord';
+
+module.exports = drawHeatmap;
+
+function drawHeatmap(painter: Painter, sourceCache: SourceCache, layer: HeatmapStyleLayer, coords: Array<TileCoord>) {
+    if (painter.isOpaquePass) return;
+
+    const gl = painter.gl;
+
+    painter.setDepthSublayer(0);
+    painter.depthMask(false);
+
+    // Allow circles to be drawn across boundaries, so that
+    // large circles are not clipped to tiles
+    gl.disable(gl.STENCIL_TEST);
+
+    renderToTexture(gl, painter, layer);
+
+    gl.clearColor(0, 0, 0, 0);
+    gl.clear(gl.COLOR_BUFFER_BIT);
+
+    gl.blendFunc(gl.ONE, gl.ONE);
```

Comment:
Would be good to add a comment here to explain the significance of this blend function, since it's essential to how the kernel density method works (and maybe some comments in general on this method)